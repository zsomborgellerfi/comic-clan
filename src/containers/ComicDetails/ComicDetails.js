import React, { Fragment } from "react";
import { connect } from "react-redux";
import Loading from "../../components/Loading";
import ComicsList from "../../components/ComicsList";

import { getComics } from "../../actions";
import { Container } from "react-bootstrap";
import _ from "lodash";
import { Redirect, Link } from "react-router-dom";
import { IoIosArrowBack } from "react-icons/io";
import fullStar from "./full_star.svg";
import emptyStar from "./empty_star.png";
import "./ComicDetails.scss";

class ComicDetails extends React.Component {
  componentDidMount() {
    this.fetchComics();
  }

  fetchComics = () => {
    this.props.getComics();
  };

  render() {
    const { loading, comics, comicsFetched } = this.props;

    // it would be more ideal to get one comic by id param
    const comic = comics.find((comic) => {
      return comic.name.includes(this.props.match.params.name);
    });

    const otherComics = _.shuffle(
      comics
        .filter((comic) => !comic.name.includes(this.props.match.params.name))
        .slice(0, 5)
    );

    return (
      <Container fluid className={"ComicDetails"}>
        <div className={"ComicDetails__link"}>
          <Link to={"/comics"}>
            <IoIosArrowBack />
            <span>Back to collection</span>
          </Link>
        </div>
        {comic && (
          <Fragment>
            <div className={"ComicDetails__comicWrapper"}>
              <div className={"ComicDetails__image"}>
                <img src={comic.image} alt={comic.name} />
              </div>
              <div className={"ComicDetails__comic"}>
                <div className={"ComicDetails__name"}>
                  <h1>
                    {" "}
                    {comic.name} ({comic.year}){" "}
                  </h1>
                  <div className={"ComicDetails__ratings"}>
                    {[...Array(comic.rating)].map(() => (
                      <img src={fullStar} alt={`${comic.rating} star`} />
                    ))}
                    {[...Array(5 - comic.rating)].map(() => (
                      <img src={emptyStar} alt={`${comic.rating} star`} />
                    ))}
                  </div>
                </div>
                <div className={"ComicDetails__info"}>
                  <div>
                    Writer: <span>{comic.writer}</span>
                  </div>
                  <div>
                    Artist: <span>{comic.artist}</span>
                  </div>
                  <div>
                    Publication: <span>{comic.publication}</span>
                  </div>
                  <div>
                    Owner: <span>{comic.owner}</span>{" "}
                  </div>
                </div>
                <div className={"ComicDetails__summary"}>{comic.summary}</div>
              </div>
            </div>
            <hr className={"separator"} />
            <h3>Other Random Books</h3>
            <ComicsList comics={otherComics} />
          </Fragment>
        )}
        {loading && <Loading />}
        {comicsFetched && !comics.length && <Redirect to={"/"} />}
      </Container>
    );
  }
}

ComicDetails.defaultProps = {
  comics: [],
};

const mapStateToProps = (state) => {
  return {
    comics: state.comics,
    loading: state.loading,
    comicsFetched: state.comicsFetched,
  };
};
const mapDispatchToProps = {
  getComics: getComics,
};

export default connect(mapStateToProps, mapDispatchToProps)(ComicDetails);
