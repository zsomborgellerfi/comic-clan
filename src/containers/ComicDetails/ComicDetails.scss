@import '../../styles/variables';

.ComicDetails {
  background: $bg-color-2;
  min-height: $full-height;
  &__comicWrapper {
    display: flex;
    flex-flow: row wrap;
    margin-bottom: 70px;
    @media #{$mq-md} {
      flex-flow: row;
      flex-direction: row;
    }
  }
  &__image {
    margin-right: 25px;
    margin-bottom: 40px;

    img {
      width: 200px;
      @media #{$mq-md} {
        width: 340px;
        margin-bottom: 0;
      }
    }
  }
  &__name {
    display: flex;
    flex-flow: row wrap;

    h1 {
      font-size: 32px;
      line-height: 37px;
      margin: 0 40px 0 0;
      color: $text-color-5;
    }
  }
  &__ratings {
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: row;
    margin-top: 10px;
    @media #{$mq-md} {
      margin-top: 0;
    }
    img {
      width: 30px;
      height: auto;
    }
  }
  &__info {
    color: $text-color-3;
    margin: 14px 0 34px;
    line-height: 20px;
    font-weight: 500;
    div {
      margin-bottom: 10px;
      span {
        color: $text-color-1;
        font-weight: bold;
        margin-left: 5px;
      }
    }
  }
  &__summary {
    color: $text-color-1;
    font-weight: bold;
  }
  &__link {
    padding: 32px 0 36px;
    a {
      color: $text-color-4;
      font-size: 20px;
      color: #777777;
      font-weight: 500;
      span {
        margin-left: 5px;
        text-decoration-line: underline;
      }
    }
  }
  h3 {
    margin: 30px 0;
    font-weight: normal;
    font-size: 32px;
    line-height: 37px;
    color: $text-color-5;
  }
}
