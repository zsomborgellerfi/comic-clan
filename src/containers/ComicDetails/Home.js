import React from "react";
import { connect } from "react-redux";
import Loading from "../../components/Loading";
import ComicsList from "../../components/ComicsList";
import { LinkContainer } from "react-router-bootstrap";

import { getComics } from "../../actions";
import "./Home.scss";
import { comicFilterKeys } from "../../utils/comicFilterKeys";
import { Form, InputGroup, Button, Container } from "react-bootstrap";
import _ from "lodash";

class Home extends React.Component {
  constructor(props) {
    super(props);
    const urlCategory = props.match.params.category;
    this.state = {
      activeFilterKey:
        urlCategory && comicFilterKeys.includes(urlCategory)
          ? urlCategory
          : comicFilterKeys[0],
      comicsGroupedByKey: {},
      searchText: "",
    };
  }

  componentWillMount() {
    this.fetchComics();
  }

  componentDidUpdate(prevProps, prevState) {
    const { comics, location, match } = this.props;
    if (prevState.searchText !== this.state.searchText) {
      this.fetchComics();
    }
    if (
      comics.length !== prevProps.comics.length ||
      location.pathname !== prevProps.location.pathname
    ) {
      this.filterComics(match.params.category || this.state.activeFilterKey);
    }
  }

  fetchComics = () => {
    this.props.getComics(this.state.searchText);
  };

  filterComics = (key) => {
    const { comics } = this.props;

    const groupBy = (key) => (array) =>
      array.reduce(
        (objectsByKeyValue, obj) => ({
          ...objectsByKeyValue,
          [obj[key]]: (objectsByKeyValue[obj[key]] || []).concat(obj),
        }),
        {}
      );

    let comicsGroupedByKey = {};
    if (key === "random") {
      comicsGroupedByKey = { [key]: _.shuffle(comics) };
    } else {
      comicsGroupedByKey = groupBy(key)(comics);
    }

    this.setState({ activeFilterKey: key, comicsGroupedByKey });
  };

  onSearchInputChange = (event) => {
    /* signal to React not to nullify the event object */
    event.persist();

    if (!this.debouncedFn) {
      this.debouncedFn = _.debounce(() => {
        let searchText = event.target.value;
        this.setState({ searchText });
      }, 300);
    }

    this.debouncedFn();
  };

  render() {
    const { loading } = this.props;
    const { comicsGroupedByKey } = this.state;

    return (
      <Container fluid className={"Home__container"}>
        <div>
          <Form onSubmit={(e) => e.preventDefault()}>
            <InputGroup className="mb-3">
              <InputGroup.Prepend>
                <InputGroup.Text id="basic-addon1">@</InputGroup.Text>
              </InputGroup.Prepend>
              <Form.Control
                placeholder="Search by book name"
                aria-label="search"
                aria-describedby="search-book"
                onChange={this.onSearchInputChange}
              />
            </InputGroup>
          </Form>
          {comicFilterKeys.map((key) => {
            return (
              <LinkContainer to={`/comics/${key}`}>
                <Button>{key}</Button>
              </LinkContainer>
            );
          })}
          {!loading ? (
            Object.keys(comicsGroupedByKey).map((key) => {
              return (
                <div key={key}>
                  <h3>{key}</h3>
                  <ComicsList comics={comicsGroupedByKey[key]} />
                </div>
              );
            })
          ) : (
            <Loading />
          )}
        </div>
      </Container>
    );
  }
}

Home.defaultProps = {
  comics: [],
};
const mapStateToProps = (state) => {
  return {
    comics: state.comics,
    loading: state.loading,
  };
};
const mapDispatchToProps = {
  getComics: getComics,
};

export default connect(mapStateToProps, mapDispatchToProps)(Home);
