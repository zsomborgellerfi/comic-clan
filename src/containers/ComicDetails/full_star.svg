<svg width="33" height="32" viewBox="0 0 33 32" fill="none" xmlns="http://www.w3.org/2000/svg">
<g filter="url(#filter0_d)">
<path d="M16.5 21.0875L8.775 25.75L10.825 16.9625L4 11.05L12.9875 10.2875L16.5 2L20.0125 10.2875L29 11.05L22.175 16.9625L24.225 25.75L16.5 21.0875Z" fill="url(#paint0_linear)"/>
</g>
<defs>
<filter id="filter0_d" x="0" y="0" width="33" height="31.75" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB">
<feFlood flood-opacity="0" result="BackgroundImageFix"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0"/>
<feOffset dy="2"/>
<feGaussianBlur stdDeviation="2"/>
<feColorMatrix type="matrix" values="0 0 0 0 0.133333 0 0 0 0 0.133333 0 0 0 0 0.133333 0 0 0 0.7 0"/>
<feBlend mode="normal" in2="BackgroundImageFix" result="effect1_dropShadow"/>
<feBlend mode="normal" in="SourceGraphic" in2="effect1_dropShadow" result="shape"/>
</filter>
<linearGradient id="paint0_linear" x1="29" y1="2" x2="29" y2="25.75" gradientUnits="userSpaceOnUse">
<stop stop-color="#F4C782"/>
<stop offset="1" stop-color="#E4964B"/>
</linearGradient>
</defs>
</svg>
