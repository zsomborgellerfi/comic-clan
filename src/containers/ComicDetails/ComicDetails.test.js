import React from "react";
import { shallow } from "enzyme";
import { ComicDetails } from "./ComicDetails";

const props = {
  getComics: jest.fn(),
  comics: [{ name: "test-name", rating: 3 }],
  match: {
    params: {
      name: "test-name",
    },
  },
};
describe("ComicDetails", () => {
  const wrapper = shallow(<ComicDetails {...props} />);

  it("should render ComicDetails correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
