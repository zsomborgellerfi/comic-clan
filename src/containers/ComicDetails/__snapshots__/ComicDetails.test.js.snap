// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComicDetails should render ComicDetails correctly 1`] = `
<Container
  className="ComicDetails"
  fluid={true}
>
  <div
    className="ComicDetails__link"
  >
    <Link
      to="/comics"
    >
      <IoIosArrowBack />
      <span>
        Back to collection
      </span>
    </Link>
  </div>
  <div
    className="ComicDetails__comicWrapper"
  >
    <div
      className="ComicDetails__image"
    >
      <img
        alt="test-name"
      />
    </div>
    <div
      className="ComicDetails__comic"
    >
      <div
        className="ComicDetails__name"
      >
        <h1>
           
          test-name
           (
          )
           
        </h1>
        <div
          className="ComicDetails__ratings"
        >
          <img
            alt="3 star"
            key="0"
            src="full_star.svg"
          />
          <img
            alt="3 star"
            key="1"
            src="full_star.svg"
          />
          <img
            alt="3 star"
            key="2"
            src="full_star.svg"
          />
          <img
            alt="3 star"
            key="0"
            src="empty_star.png"
          />
          <img
            alt="3 star"
            key="1"
            src="empty_star.png"
          />
        </div>
      </div>
      <div
        className="ComicDetails__info"
      >
        <div>
          Writer: 
          <span />
        </div>
        <div>
          Artist: 
          <span />
        </div>
        <div>
          Publication: 
          <span />
        </div>
        <div>
          Owner: 
          <span />
           
        </div>
      </div>
      <div
        className="ComicDetails__summary"
      />
    </div>
  </div>
  <hr
    className="separator"
  />
  <h3>
    Other Random Books
  </h3>
  <ComicsList
    comics={Array []}
  />
</Container>
`;
